<a href="https://colab.research.google.com/github/ManishKr19/Edureka-Data-Science/blob/main/M7_CS2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
## Loading dependancy
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
plt.rcParams['axes.labelsize'] = 14
plt.rcParams['xtick.labelsize'] = 12
plt.rcParams['ytick.labelsize'] = 12
%matplotlib inline
# to make this notebook's output stable across runs
np.random.seed(42)

In [2]:
# Loading the dataset
horseDf = pd.read_csv(r"/content/horse.csv")

In [3]:
horseDf

surgery    age  hospital_number  ...  lesion_2  lesion_3  cp_data
0        no  adult           530101  ...         0         0       no
1       yes  adult           534817  ...         0         0       no
2        no  adult           530334  ...         0         0      yes
3       yes  young          5290409  ...         0         0      yes
4        no  adult           530255  ...         0         0       no
..      ...    ...              ...  ...       ...       ...      ...
294     yes  adult           533886  ...         0         0       no
295      no  adult           527702  ...         0         0      yes
296     yes  adult           529386  ...         0         0       no
297     yes  adult           530612  ...         0         0      yes
298     yes  adult           534618  ...         0         0       no

[299 rows x 28 columns]

In [4]:
horseDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   surgery                299 non-null    object 
 1   age                    299 non-null    object 
 2   hospital_number        299 non-null    int64  
 3   rectal_temp            239 non-null    float64
 4   pulse                  275 non-null    float64
 5   respiratory_rate       241 non-null    float64
 6   temp_of_extremities    243 non-null    object 
 7   peripheral_pulse       230 non-null    object 
 8   mucous_membrane        252 non-null    object 
 9   capillary_refill_time  267 non-null    object 
 10  pain                   244 non-null    object 
 11  peristalsis            255 non-null    object 
 12  abdominal_distention   243 non-null    object 
 13  nasogastric_tube       195 non-null    object 
 14  nasogastric_reflux     193 non-null    object 
 15  nasoga

### Q1: Info command on data frame can be used to check for nulls. We can see there are lot of fields where number of records is less than 299 and all this because they have null

In [5]:
np.sum(horseDf.isnull())

surgery                    0
age                        0
hospital_number            0
rectal_temp               60
pulse                     24
respiratory_rate          58
temp_of_extremities       56
peripheral_pulse          69
mucous_membrane           47
capillary_refill_time     32
pain                      55
peristalsis               44
abdominal_distention      56
nasogastric_tube         104
nasogastric_reflux       106
nasogastric_reflux_ph    246
rectal_exam_feces        102
abdomen                  118
packed_cell_volume        29
total_protein             33
abdomo_appearance        165
abdomo_protein           198
outcome                    0
surgical_lesion            0
lesion_1                   0
lesion_2                   0
lesion_3                   0
cp_data                    0
dtype: int64

In [6]:
incomplete_rows = horseDf[horseDf.isnull().any(axis=1)]
incomplete_rows.shape

(293, 28)

In [7]:
horseDf

categorical_columns = list(horseDf.select_dtypes(include='object').columns)

categorical_columns_indexes = [horseDf.columns.get_loc(column) for column in categorical_columns]


### 2. a. Attempting the categorical encoding using pandas

In [8]:
dummyHorse = horseDf.copy()

In [9]:
for column in categorical_columns:
    print(f"++++++{column}+++++++++")
    print(dummyHorse[column].value_counts())

++++++surgery+++++++++
yes    180
no     119
Name: surgery, dtype: int64
++++++age+++++++++
adult    275
young     24
Name: age, dtype: int64
++++++temp_of_extremities+++++++++
cool      108
normal     78
warm       30
cold       27
Name: temp_of_extremities, dtype: int64
++++++peripheral_pulse+++++++++
normal       114
reduced      103
absent         8
increased      5
Name: peripheral_pulse, dtype: int64
++++++mucous_membrane+++++++++
normal_pink      79
pale_pink        57
pale_cyanotic    41
bright_pink      30
bright_red       25
dark_cyanotic    20
Name: mucous_membrane, dtype: int64
++++++capillary_refill_time+++++++++
less_3_sec    187
more_3_sec     78
3               2
Name: capillary_refill_time, dtype: int64
++++++pain+++++++++
mild_pain       67
depressed       59
extreme_pain    42
severe_pain     38
alert           38
Name: pain, dtype: int64
++++++peristalsis+++++++++
hypomotile     127
absent          73
hypermotile     39
normal          16
Name: peristalsis, dtype: i

In [10]:
dummyHorseT = pd.get_dummies(dummyHorse, prefix = categorical_columns)
dummyHorse.columns

Index(['surgery', 'age', 'hospital_number', 'rectal_temp', 'pulse',
       'respiratory_rate', 'temp_of_extremities', 'peripheral_pulse',
       'mucous_membrane', 'capillary_refill_time', 'pain', 'peristalsis',
       'abdominal_distention', 'nasogastric_tube', 'nasogastric_reflux',
       'nasogastric_reflux_ph', 'rectal_exam_feces', 'abdomen',
       'packed_cell_volume', 'total_protein', 'abdomo_appearance',
       'abdomo_protein', 'outcome', 'surgical_lesion', 'lesion_1', 'lesion_2',
       'lesion_3', 'cp_data'],
      dtype='object')

In [11]:
dummyHorseT.columns

Index(['hospital_number', 'rectal_temp', 'pulse', 'respiratory_rate',
       'nasogastric_reflux_ph', 'packed_cell_volume', 'total_protein',
       'abdomo_protein', 'lesion_1', 'lesion_2', 'lesion_3', 'surgery_no',
       'surgery_yes', 'age_adult', 'age_young', 'temp_of_extremities_cold',
       'temp_of_extremities_cool', 'temp_of_extremities_normal',
       'temp_of_extremities_warm', 'peripheral_pulse_absent',
       'peripheral_pulse_increased', 'peripheral_pulse_normal',
       'peripheral_pulse_reduced', 'mucous_membrane_bright_pink',
       'mucous_membrane_bright_red', 'mucous_membrane_dark_cyanotic',
       'mucous_membrane_normal_pink', 'mucous_membrane_pale_cyanotic',
       'mucous_membrane_pale_pink', 'capillary_refill_time_3',
       'capillary_refill_time_less_3_sec', 'capillary_refill_time_more_3_sec',
       'pain_alert', 'pain_depressed', 'pain_extreme_pain', 'pain_mild_pain',
       'pain_severe_pain', 'peristalsis_absent', 'peristalsis_hypermotile',
       'perist

In [12]:
print(dummyHorseT.shape)
print(horseDf.shape)
horseDf.info()

(299, 70)
(299, 28)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   surgery                299 non-null    object 
 1   age                    299 non-null    object 
 2   hospital_number        299 non-null    int64  
 3   rectal_temp            239 non-null    float64
 4   pulse                  275 non-null    float64
 5   respiratory_rate       241 non-null    float64
 6   temp_of_extremities    243 non-null    object 
 7   peripheral_pulse       230 non-null    object 
 8   mucous_membrane        252 non-null    object 
 9   capillary_refill_time  267 non-null    object 
 10  pain                   244 non-null    object 
 11  peristalsis            255 non-null    object 
 12  abdominal_distention   243 non-null    object 
 13  nasogastric_tube       195 non-null    object 
 14  nasogastric_reflux     193 non-null   

In [13]:
dummyHorseT.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 70 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   hospital_number                   299 non-null    int64  
 1   rectal_temp                       239 non-null    float64
 2   pulse                             275 non-null    float64
 3   respiratory_rate                  241 non-null    float64
 4   nasogastric_reflux_ph             53 non-null     float64
 5   packed_cell_volume                270 non-null    float64
 6   total_protein                     266 non-null    float64
 7   abdomo_protein                    101 non-null    float64
 8   lesion_1                          299 non-null    int64  
 9   lesion_2                          299 non-null    int64  
 10  lesion_3                          299 non-null    int64  
 11  surgery_no                        299 non-null    uint8  
 12  surgery_

Implementing Imputers

In [19]:
from sklearn.impute import SimpleImputer

In [21]:
imputer = SimpleImputer(strategy="most_frequent")

In [22]:
X = imputer.fit_transform(dummyHorseT)

In [23]:
dummyHorseWN = pd.DataFrame(X, columns = dummyHorseT.columns, index = list(horseDf.index.values))

In [24]:
dummyHorseWN.loc[incomplete_rows.index.values].head()

hospital_number  rectal_temp  ...  cp_data_no  cp_data_yes
0         530101.0         38.5  ...         1.0          0.0
1         534817.0         39.2  ...         1.0          0.0
2         530334.0         38.3  ...         0.0          1.0
3        5290409.0         39.1  ...         0.0          1.0
4         530255.0         37.3  ...         1.0          0.0

[5 rows x 70 columns]

### Fit a decision Tree and check for accuracy

In [25]:
horseDf.head()

surgery    age  hospital_number  ...  lesion_2  lesion_3  cp_data
0      no  adult           530101  ...         0         0       no
1     yes  adult           534817  ...         0         0       no
2      no  adult           530334  ...         0         0      yes
3     yes  young          5290409  ...         0         0      yes
4      no  adult           530255  ...         0         0       no

[5 rows x 28 columns]